src/megaclient.cpp

```diff
@@ -2463,16 +2471,64 @@ bool MegaClient::dispatch(direction_t d)
             app->transfer_prepare(nexttransfer);
         }
 
+        bool openok;
+        bool openfinished = false;
+
         // verify that a local path was given and start/resume transfer
         if (nexttransfer->localfilename.size())
         {
-            // allocate transfer slot
-            ts = new TransferSlot(nexttransfer);
+            if (!nexttransfer->slot)
+            {
+                // allocate transfer slot
+                ts = new TransferSlot(nexttransfer);
+            }
+            else
+            {
+                ts = nexttransfer->slot;
+            }
+
+            if (ts->fa->asyncavailable())
+            {
+                if (!nexttransfer->asyncopencontext)
+                {
+                    LOG_debug << "Starting async open";
+                    nexttransfer->asyncopencontext = (d == PUT)
+                        ? ts->fa->asyncfopen(&nexttransfer->localfilename, true, false)
+                        : ts->fa->asyncfopen(&nexttransfer->localfilename, false, true, nexttransfer->size);
+                    asyncfopens++;
+                }
+
```

Comment:
Yes, really asynchronous file opens are not fully supported yet. I have added a comment below to show what's missing. Anyway, all filesystem layers in the SDK always finishes asyncfopen synchronously.